### ISSUE TITLE
Proposal: multiline completion support

### ISSUE THREAD

**Issue by hauleth** [vim/vim#2505]
Currently Vim allows only single line completions. This is quite limiting as there is also no way to determine beginning and end of the completed text so this leaves us in the limbo where we need to do some hackish things to achieve things like multiline snippets.

Most obvious way to do such thing would be to make (`:h complete-items`):

```viml
{ 'word': "aaa\nbbb" }
```

to insert 2 lines:

```
aaa
bbb
```

Currently it inserts

```
aaa^@bbb
```

Where `^@` is NUL character (`:h NL-used-for-NUL`).

The problems with multiline completion is that there will be open question how to handle indentation.

Alternatively we could allow `List` as `'word'` parameter where each line would be one item.

**Comment by inkarkat** [comment]
I would welcome this; the current behavior of inserting `^@` presumably is just an implementation artifact, not useful behavior. In fact, I've been emulating this (for user completions) for a long time with the following snippet. That works fine, but having this built-in would be even better. As long as the only source for multi-line completions is user completions, I don't see an issue with adjusting indentation; the completion function could do that, no need for the completion itself to interfere.

    "			Fix the multi-line completion problem.
    "			Currently, completion matches are inserted literally,
    "			with newlines represented by ^@. Vim should expand this
    "			into proper newlines. We work around this via an autocmd
    "			that fires once after completion, and by hooking into
    "			the <CR> key. (<C-Y> apparently works even without
    "			this.)
    "			(FIXME: It seems to be fired once after the initial
    "			completion trigger with empty 'completeopt'.)
    "			Usage: All CTRL-X_... completion mappings that may
    "			return multi-line matches must append
    "			<Plug>(CompleteMultilineFixSetup) to get this fix.
    function! s:CompleteMultilineFix()
        let l:textBeforeCursor = strpart(getline('.'), 0, col('.') - 1)
        let l:lastNewlineCol = strridx(l:textBeforeCursor, "\n")
        if l:lastNewlineCol == -1
            " Nothing to do.
            return ''
        endif

        execute "substitute/\n/\\r/ge"

        " The substitute command positions the cursor at the first column of the
        " last line inserted. This is fine when the completion ended with a newline,
        " but needs correction when an incomplete last line has been inserted.
        call cursor(line('.'), len(l:textBeforeCursor) - l:lastNewlineCol)

        call histdel('search', -1)

        return ''
    endfunction
    function! s:CompleteMultilineFixSetup()
        augroup CompleteMultilineFix
            autocmd! CompleteDone * call <SID>CompleteMultilineFix() | autocmd! CompleteMultilineFix
        augroup END

        return ''
    endfunction
    inoremap <expr> <Plug>(CompleteMultilineFixSetup) <SID>CompleteMultilineFixSetup()

**Comment by hauleth** [comment]
@inkarkat I also have tried that with my [usnip.vim](https://github.com/hauleth/usnip.vim) plugin, but the problem is that when you have `set fo=t` it will fail when your insertion is longer than `&tw`.

Also it would be handy to have way to determine the start and the end of the completed text in`CompleteDone` auto command, which would become handy for supporting LSP snippets syntax.

**Comment by chrisbra** [comment]
> Alternatively we could allow List as 'word' parameter where each line would be one item.

+1 for this

**Comment by euclio** [comment]
Perhaps this could be generalized: completions may insert any number of strings at arbitrary positions. For example, Java autocomplete might like to add the relevant import for a class when inserting the name. In fact, the Java language server already reports such additional edits.

**Comment by hauleth** [comment]
@euclio I think that this would be a little bit too much. If you want to launch command to add imports in Java you can already do so using `CompleteDone` auto command. This one shouldn't be hard. On the other hand it would be handy to have some way to add arbitrary data to `v:completed_item`, like `meta` field, so it would provide additional information for use in `CompleteDone` auto command.

**Comment by puremourning** [comment]
> v:completed_item, like meta field, so it would provide additional information for use in CompleteDone auto command. 

This. This would also allow, for example, a snippet to be expanded after insertion. I actually wrote the code to do this a few years ago but didn't have the courage to raise a PR. If people are still interested in this, them let me know. The idea was to add a user dict to each completion item, which can be inspected in `CompleteDone` via the `v:completd_item`

**Comment by hauleth** [comment]
@puremourning snippets system is main reason why I have issued this proposal.

I think that there should be added 2 another:

- marks/variables for beginning and end of completed text that would be usable inside `CompleteDone` auto command
- additional `meta` field in `completion-items` that would allow providing additional information for `CompleteDone` auto command

**Comment by puremourning** [comment]
i agree. that makes a lot of sense. YCM has a load of annoying code that tries to guess the location of the just-completed-text and i've seen other plugins/scripts doing similar sorts of things.

**Comment by GavinHsueh** [comment]
good idea～

**Comment by brammool** [comment]
Łukasz Jan Niemier wrote:

> Currently Vim allows only single line completions. This is quite
> limiting as there is also no way to determine beginning and end of the
> completed text so this leaves us in the limbo where we need to do some
> hackish things to achieve things like multiline snippets.
> 
> Most obvious way to do such thing would be to make (`:h complete-items`):
> 
> ```viml
> { 'word': "aaa\nbbb" }
> ```
> 
> to insert 2 lines:
> 
> ```
> aaa
> bbb
> ```
> 
> Currently it inserts
> 
> ```
> aaa^@bbb
> ```
> 
> Where `^@` is NUL character (`:h NL-used-for-NUL`).
> 
> The problems with multiline completion is that there will be open question how to handle indentation.

The usual way to deal with this is "as if you typed the text".  That's
probably also the simplest way to implement it.

> Alternatively we could allow `List` as `'word'` parameter where each line would be one item.

Inserting a NUL character is rare enough we can support using \n here.

Anyway, before making changes to Insert mode completion, this code is
long overdue for a refactoring.  There are a few bugs and it's using
very confusing variables to keep the current state.  First step would be
to properly define what states there are.  That's not easy, since the
search for completions continues when already showing the first matches.

-- 
hundred-and-one symptoms of being an internet addict:
203. You're an active member of more than 20 newsgroups.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
